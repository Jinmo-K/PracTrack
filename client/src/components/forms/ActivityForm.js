import React, { useState } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
// Components
import { ModalHeader, ModalBody, ModalFooter, Button } from 'reactstrap';
import TextField from '@material-ui/core/TextField';
// Functions
import { useInput } from '../../hooks/useInput';
import { hideForm } from '../../actions/formActions';
import { addActivity } from '../../actions/activitiesActions';


/** 
 *   New activity form, displayed in AppModal
 */
const ActivityForm = ({ hideForm, addActivity, userId, activities }) => {
  const { value: name, bindProps: bindName } = useInput('');
  const [nameError, setNameError] = useState('');
  const [goal, setGoal] = useState('');
  const [goalError, setGoalError] = useState('');

  /**
   * Validates name and sets name error if input is empty or activity name already exists
   * @return {boolean}  True if name is valid
   */
  const validateName = () => {
    let error = '';
    if (!name) {
      error = 'Activity must have a name';
    }
    else if (activities.some(activity => activity.title.toLowerCase() === name.toLowerCase())) {
      error = 'Activities must have unique names';
    }
    setNameError(error);
    return !error;
  }

  /**
   * Handles submit button click. If name is valid, dispatches functions to add 
   * the new activity and close the form modal.
   */ 
  const onSubmit = (e) => {
    e.preventDefault();
    let valid = validateName();
    if (valid) {
      const newActivity = {
        title: name,
        goal: goal
      };
      addActivity(userId, newActivity);
      hideForm();
    }
  };

  return (
    <div>
      <ModalHeader toggle={hideForm}>
        Create a new activity
      </ModalHeader>
      <ModalBody>
        <div id='activity-form' className='p-0 p-sm-2 justify-content-center container'>
          {/* Name */}
          <div className='form-group row mb-1 mb-sm-3 align-items-center'>
            <label className='col-12 col-sm-2 text-left h1 lead text-sm-right' htmlFor='inputName'>
              Name
            </label>
            <div className='col-12 col-sm-10'>
              <TextField
                id='inputName'
                variant="outlined"
                size='small'
                fullWidth
                error={!!nameError}
                helperText={nameError}
                onFocus={() => setNameError('')}
                {...bindName}
              />
            </div>
          </div>

          {/* Goal */}
          <div className='form-group row mb-1 mb-sm-3 align-items-center'>
            <label className='col-12 col-sm-2 text-left h1 lead text-sm-right' htmlFor='inputGoal'>
              Goal
            </label>
            <div className='col-12 col-sm-10'>
              <TextField
                id='inputGoal'
                variant="outlined"
                size='small'
                fullWidth
                type='number'
                error={!!goalError}
                helperText={goalError}
                InputProps={{
                  inputProps: { min: 0 }
                }}
                value={goal}
                onChange={(e) => {
                  (e.target.value < 0) ? setGoalError('Goal cannot be negative') : setGoalError('');
                  setGoal(e.target.value);
                }}
              />
            </div>
          </div>
        </div>
      </ModalBody>
      <ModalFooter>
        <Button id='activity-form-cancel-btn' color="secondary" onClick={hideForm}>Cancel</Button>{' '}
        <Button id='activity-form-submit-btn' color="info" onClick={onSubmit} disabled={!!goalError}>Submit</Button>
      </ModalFooter>
    </div>
  );
}

ActivityForm.propTypes = {
  userId: PropTypes.string.isRequired,
  hideForm: PropTypes.func.isRequired,
  addActivity: PropTypes.func.isRequired,
  activities: PropTypes.array.isRequired,
}

const mapStateToProps = (state) => ({
  userId: state.auth.user.id,
  activities: state.activities,
});

const mapDispatchToProps = {
  hideForm,
  addActivity,
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ActivityForm);

export { ActivityForm };