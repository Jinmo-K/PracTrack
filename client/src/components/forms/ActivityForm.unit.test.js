import React from 'react';
import { mount, shallow } from 'enzyme';
import { ActivityForm } from './ActivityForm';

describe('<ActivityForm />', () => {
  let wrapper;
  const mockEvent = {
    preventDefault: jest.fn(),
  };
  const props= {
    hideForm: jest.fn(),
    addActivity: jest.fn(),
    userId: '0',
    activities: [
      { title: 'invalid' }
    ]
  };

  beforeEach(() => {
    jest.clearAllMocks();
    wrapper = shallow(<ActivityForm {...props} />);
  });

  it('Should render initial layout given required props', () => {
    expect(wrapper.getElements()).toMatchSnapshot();
  });

  it('Should reflect user changes in name input', () => {
    wrapper.find('#inputName').simulate('change', { target: { value: 'valid' }});
    expect(wrapper.find('#inputName').props().value).toEqual('valid');
  });

  it('Should reflect user changes in goal input', () => {
    wrapper.find('#inputGoal').simulate('change', { target: { value: 10 }});
    expect(wrapper.find('#inputGoal').props().value).toEqual(10);
  });

  it('Should call hideForm when \'Cancel\' button is clicked', () => {
    const cancelBtn = wrapper.find('#activity-form-cancel-btn');
    cancelBtn.simulate('click');
    expect(props.hideForm).toBeCalledTimes(1);
  });

  it('Should call hideForm when modal header close button is clicked', () => {
    // Mount to render ModalHeader
    wrapper = mount(<ActivityForm {...props} />);
    const closeBtn = wrapper.find('button.close');
    closeBtn.simulate('click');
    expect(props.hideForm).toBeCalledTimes(1);
  });

  it('Should submit and close the form with valid name', () => {
    wrapper.find('#inputName').simulate('change', { target: { value: 'valid' }});
    const submitBtn = wrapper.find('#activity-form-submit-btn');
    submitBtn.simulate('click', mockEvent);
    expect(props.addActivity).toBeCalledTimes(1);
    expect(props.hideForm).toBeCalledTimes(1);
  });

  it('Should not submit or close the form with duplicate name', () => {
    wrapper.find('#inputName').simulate('change', { target: { value: 'invalid' }});
    const submitBtn = wrapper.find('#activity-form-submit-btn');
    submitBtn.simulate('click', mockEvent);
    expect(props.addActivity).toBeCalledTimes(0);
    expect(props.hideForm).toBeCalledTimes(0);
  });

  it('Should not submit or close the form with empty name', () => {
    wrapper.find('#inputName').simulate('change', { target: { value: '' }});
    const submitBtn = wrapper.find('#activity-form-submit-btn');
    submitBtn.simulate('click', mockEvent);
    expect(props.addActivity).toBeCalledTimes(0);
    expect(props.hideForm).toBeCalledTimes(0);
  });

  it('Should not render any name error on submit with valid input', () => {
    wrapper.find('#inputName').simulate('change', { target: { value: 'valid' }});
    const submitBtn = wrapper.find('#activity-form-submit-btn');
    submitBtn.simulate('click', mockEvent);
    expect(wrapper.find('#inputName').props().error).toEqual(false);
    expect(wrapper.find('#inputName').props().helperText).toEqual('');
  });

  it('Should render duplicate name error on submit with duplicate name', () => {
    wrapper.find('#inputName').simulate('change', { target: { value: 'invalid' }});
    const submitBtn = wrapper.find('#activity-form-submit-btn');
    submitBtn.simulate('click', mockEvent);
    expect(wrapper.find('#inputName').props().error).toEqual(true);
    expect(wrapper.find('#inputName').props().helperText).toEqual('Activities must have unique names');
  });

  it('Should render empty name error on submit with empty name', () => {
    wrapper.find('#inputName').simulate('change', { target: { value: '' }});
    const submitBtn = wrapper.find('#activity-form-submit-btn');
    submitBtn.simulate('click', mockEvent);
    expect(wrapper.find('#inputName').props().error).toEqual(true);
    expect(wrapper.find('#inputName').props().helperText).toEqual('Activity must have a name');
  });

  it('Should reset name error on name input focus', () => {
    wrapper.find('#inputName').simulate('change', { target: { value: 'invalid' }});
    const submitBtn = wrapper.find('#activity-form-submit-btn');
    submitBtn.simulate('click', mockEvent);
    expect(wrapper.find('#inputName').props().error).toEqual(true);
    expect(wrapper.find('#inputName').props().helperText).toEqual('Activities must have unique names');
    wrapper.find('#inputName').simulate('focus');
    expect(wrapper.find('#inputName').props().error).toEqual(false);
    expect(wrapper.find('#inputName').props().helperText).toEqual('');
  });

  it('Should render goal error on negative goal input', () => {
    wrapper.find('#inputGoal').simulate('change', { target: { value: -10 }});
    expect(wrapper.find('#inputGoal').props().error).toEqual(true);
    expect(wrapper.find('#inputGoal').props().helperText).toEqual('Goal cannot be negative');
  });

  it('Should not have goal error on valid goal input', () => {
    wrapper.find('#inputGoal').simulate('change', { target: { value: 10 }});
    expect(wrapper.find('#inputGoal').props().error).toEqual(false);
    expect(wrapper.find('#inputGoal').props().helperText).toEqual('');
  });

  it('Should not disable submit button on valid goal input', () => {
    wrapper.find('#inputGoal').simulate('change', { target: { value: 10 }});
    expect(wrapper.find('#activity-form-submit-btn').prop('disabled')).toEqual(false);
  });

  it('Should disable submit button on negative goal input', () => {
    wrapper.find('#inputGoal').simulate('change', { target: { value: -10 }});
    expect(wrapper.find('#activity-form-submit-btn').prop('disabled')).toEqual(true);
  });
});