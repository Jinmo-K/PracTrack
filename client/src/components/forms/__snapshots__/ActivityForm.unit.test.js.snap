// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ActivityForm /> Should render initial layout given required props 1`] = `
Array [
  <div>
    <ModalHeader
      charCode={215}
      closeAriaLabel="Close"
      tag="h5"
      toggle={[MockFunction]}
      wrapTag="div"
    >
      Create a new activity
    </ModalHeader>
    <ModalBody
      tag="div"
    >
      <div
        className="p-0 p-sm-2 justify-content-center container"
        id="activity-form"
      >
        <div
          className="form-group row mb-1 mb-sm-3 align-items-center"
        >
          <label
            className="col-12 col-sm-2 text-left h1 lead text-sm-right"
            htmlFor="inputName"
          >
            Name
          </label>
          <div
            className="col-12 col-sm-10"
          >
            <ForwardRef(WithStyles)
              error={false}
              fullWidth={true}
              helperText=""
              id="inputName"
              onChange={[Function]}
              onFocus={[Function]}
              size="small"
              value=""
              variant="outlined"
            />
          </div>
        </div>
        <div
          className="form-group row mb-1 mb-sm-3 align-items-center"
        >
          <label
            className="col-12 col-sm-2 text-left h1 lead text-sm-right"
            htmlFor="inputGoal"
          >
            Goal
          </label>
          <div
            className="col-12 col-sm-10"
          >
            <ForwardRef(WithStyles)
              InputProps={
                Object {
                  "inputProps": Object {
                    "min": 0,
                  },
                }
              }
              error={false}
              fullWidth={true}
              helperText=""
              id="inputGoal"
              onChange={[Function]}
              size="small"
              type="number"
              value=""
              variant="outlined"
            />
          </div>
        </div>
      </div>
    </ModalBody>
    <ModalFooter
      tag="div"
    >
      <Button
        color="secondary"
        id="activity-form-cancel-btn"
        onClick={[MockFunction]}
        tag="button"
      >
        Cancel
      </Button>
       
      <Button
        color="info"
        disabled={false}
        id="activity-form-submit-btn"
        onClick={[Function]}
        tag="button"
      >
        Submit
      </Button>
    </ModalFooter>
  </div>,
]
`;
